core/src/main/java/com/linecorp/armeria/server/WebSocketSessionHandler.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2023 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server;
+
+import com.linecorp.armeria.common.ClosedSessionException;
+import com.linecorp.armeria.common.ContentTooLargeException;
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.internal.common.InitiateConnectionShutdown;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelDuplexHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+import io.netty.handler.codec.http.HttpContent;
+import io.netty.handler.codec.http.HttpResponse;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import io.netty.handler.codec.http.HttpServerCodec;
+import io.netty.handler.codec.http2.Http2Error;
+import io.netty.util.ReferenceCountUtil;
+
+final class WebSocketSessionHandler extends ChannelDuplexHandler {
+
+    private final StreamingDecodedHttpRequest req;
+    private final ServerHttpObjectEncoder encoder;
+    private final ServiceConfig serviceConfig;
+
+    WebSocketSessionHandler(StreamingDecodedHttpRequest req, ServerHttpObjectEncoder encoder,
+                            ServiceConfig serviceConfig) {
+        this.req = req;
+        this.encoder = encoder;
+        this.serviceConfig = serviceConfig;
+    }
+
+    @Override
+    public void channelUnregistered(ChannelHandlerContext ctx) throws Exception {
+        super.channelUnregistered(ctx);
+        req.close(ClosedSessionException.get());
+    }
+
+    @Override
+    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+        if (evt instanceof InitiateConnectionShutdown) {
+            encoder.keepAliveHandler().disconnectWhenFinished();
+            return;
+        }
+        ctx.fireUserEventTriggered(evt);
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) {
+        if (msg instanceof DecodedHttpRequest) {
+            ctx.fireChannelRead(msg);
+            return;
+        }
```

Comment:
Is there any handler next to `WebSocketSessionHandler`? If not, we shouldn't do this.